src/Interpreters/Context.cpp

```diff
@@ -3004,8 +3013,18 @@ void Context::reloadClusterConfig() const
     }
 }
 
+std::map<String, ClusterPtr> Context::getClusters() const
+{
+    auto clusters = getClustersImpl()->getContainer();
+    if (shared->cluster_discovery)
+    {
+        for (const auto & [name, cluster] : shared->cluster_discovery->getClusters())
```

Comment:
>  It can be updated anytime (on the server added to/removed from the cluster).

Yes, but that's a problem when iterating no?
Each loop iteration you will end up with different possible iterators if map changes.

Reply:
I assumed that
```cpp
for (const auto & [name, cluster] : shared->cluster_discovery->getClusters())
{
...
```

is the same as

```cpp
const auto & tmp = shared->cluster_discovery->getClusters();
for (const auto & [name, cluster] : tmp)
{
...
```

and `getClusters` is just called once.


https://godbolt.org/z/sj6Ydc7zT

But we can change the code to be like the second snippet to ensure it.

```suggestion
        const auto & cluster_discovery_map = shared->cluster_discovery->getClusters();
        for (const auto & [name, cluster] : cluster_discovery_map)
```

Reply:
there is a cool tool for things like this, C++ Insights
your [example](https://cppinsights.io/lnk?code=I2luY2x1ZGUgPGlvc3RyZWFtPgojaW5jbHVkZSA8dmVjdG9yPgoKCmNsYXNzIEZvbwp7CnB1YmxpYzoKICAgIHN0ZDo6dmVjdG9yPGludD4geCA9IHsxLDIsM307CgogICAgc3RkOjp2ZWN0b3I8aW50PiAmIGdldFgoKQogICAgewogICAgICAgIHN0ZDo6Y291dCA8PCAiRm9vOjpnZXRYXG4iOwogICAgICAgIHgucHVzaF9iYWNrKDQpOwogICAgICAgIHJldHVybiB4OwogICAgfQp9OwoKaW50IG1haW4oKQp7CiAgICBGb28gZjsKICAgIGZvciAoY29uc3QgYXV0byAmIGUgOiBmLmdldFgoKSkKICAgIHsKICAgICAgICBzdGQ6OmNvdXQgPDwgZSA8PCAiICwgIjsKICAgIH0KICAgIHN0ZDo6Y291dCA8PCAiXG4iOwp9Cg==&std=cpp2a&rev=1.0)

But yeah, I'm still afraid to have code that relies on such assumptions because it's not obvious